File: src/policy_platform/api/actor_role.py
"""Refusals that turn on who is acting, as codes rather than sentences.

Two routers each carried their own copy of one sentence:

    "Only a Policy Manager can perform this action. Switch your acting role in
     the header."

and the interface carried a third, which had already drifted from both -- it
says "launch a new campaign" where the servers say "perform this action". Three
producers of one message is three places to edit for a copy change, two of
which need a deployment, and the drift had already happened without anyone
choosing it.

WHAT A CODE HAS TO CARRY. The reason the servers wrote a sentence is that the
sentence knows something the bare refusal does not: which role was needed, and
what the actor was trying to do. A code saying only "forbidden" would leave the
interface to guess the verb, and it would guess wrong on the page that already
words it differently. So the refusal carries the role it wanted and the action
it refused, both as identifiers, and the words for each live in
`apps/web/src/actorRole.ts` beside everything else a reader sees.

The two tuples below are declared explicitly rather than derived, for the same
reason `CONDITION_PROVENANCE_CODES` is: a test enumerates them and fails when
one gains a member the interface has no wording for. A set that cannot be
enumerated cannot be checked, and an unchecked vocabulary is how a code reaches
a reader as a raw identifier.
"""
from __future__ import annotations

from typing import Final

from fastapi import HTTPException

#: The one refusal shape this module raises. Named so the interface can tell it
#: from a 403 that means something else, without matching on prose.
ACTOR_ROLE_REFUSAL: Final[str] = "actor_role_insufficient"

#: Every role a refusal can name.
ACTOR_ROLES: Final[tuple[str, ...]] = ("policy_manager",)

#: Every action a refusal can name. An action is here because a reader needs
#: the verb: "you cannot do this" and "you cannot launch a campaign" are
#: different sentences, and only the second is worth reading.
GUARDED_ACTIONS: Final[tuple[str, ...]] = ("launch_attestation_campaign",)


def require_role(actor_role: str, *, required: str, action: str) -> None:
    """Refuse, in codes, when the acting role is not the one required.

    Raises a 403 whose `detail` is an object rather than a string. Every
    consumer of this API reads `detail`, so the shape change is visible; that
    is deliberate. A refusal that silently kept a `detail` string would have
    left the old sentence in place for anything that had not been updated, and
    the drift this replaces began exactly that way.
    """

    assert required in ACTOR_ROLES, f"{required!r} is not a declared role"
    assert action in GUARDED_ACTIONS, f"{action!r} is not a declared action"

    if actor_role == required:
        return

    raise HTTPException(
        status_code=403,
        detail={
            "code": ACTOR_ROLE_REFUSAL,
            "required_role": required,
            "action": action,
        },
    )

File: apps/web/src/actorRole.ts
/**
 * The words for a refusal that turns on who is acting.
 *
 * The server used to send the sentence. Two routers each held a copy of it and
 * this interface held a third, which had already drifted: the servers said
 * "perform this action" and the page said "launch a new campaign". A copy
 * change meant editing three places, two of them behind a deployment.
 *
 * The server now sends a refusal code with the role it wanted and the action it
 * refused. Both are needed. A bare refusal code would leave this file to guess
 * the verb, and the page that already worded it differently is the proof that a
 * guess drifts.
 *
 * TWO CODES ARRIVE HERE, not one. `actor_role_insufficient` is the original,
 * raised for the named guarded actions in `api/actor_role.py`.
 * `rbac_insufficient` is raised by the capability layer in `api/authz.py`, which
 * governs every operation rather than a named few. They are recognised together
 * because a reader does not care which server module refused them, and because
 * recognising only the first would send the second down the generic-error path
 * to be rendered as a raw object — the exact failure the codes were introduced
 * to prevent.
 *
 * Their `action` fields differ in kind, and that is why the sentence below is
 * built from the role first: the original sends a named action with a verb
 * phrase written for it here, while the capability layer sends a route
 * (`POST /api/policy-sets/{key}/publish`), which has no reading for a person.
 * Naming no verb is better than naming a route at a reader.
 *
 * Which roles and actions can arrive is checked from the other side of the
 * boundary by `tests/unit/test_actor_role_wording.py`, which reads the tuples
 * out of `api/actor_role.py` and `api/roles.py` and fails when one of them has
 * no entry here. That direction cannot run from this side: the declaration is
 * Python.
 */

/** How each role is named to a reader.
 *
 *  Both vocabularies live here. `policy_manager` is the original guarded role;
 *  the three below it are the capability layer's, and a reader can be refused
 *  on either.
 */
export const ACTOR_ROLE_LABEL: Record<string, string> = {
  policy_manager: "Policy Manager",
  viewer: "Viewer",
  policy_author: "Policy Author",
  admin: "Admin",
};

/**
 * The verb phrase for each guarded action, written to sit inside "Only a
 * Policy Manager can ___".
 */
export const GUARDED_ACTION_PHRASE: Record<string, string> = {
  launch_attestation_campaign: "launch a new campaign",
};

/** What the server sends when it refuses on the acting role. */
export interface ActorRoleRefusal {
  code: string;
  required_role: string;
  action: string;
}

export const ACTOR_ROLE_REFUSAL = "actor_role_insufficient";

/** The capability layer's refusal. Same shape, different origin. */
export const RBAC_REFUSAL = "rbac_insufficient";

const REFUSAL_CODES: readonly string[] = [ACTOR_ROLE_REFUSAL, RBAC_REFUSAL];

/**
 * True when this is a refusal about who is acting rather than any other 403.
 *
 * Checked on the code, never on the words, so that rewording this file cannot
 * change which errors are recognised.
 */
export function isActorRoleRefusal(detail: unknown): detail is ActorRoleRefusal {
  return (
    typeof detail === "object" &&
    detail !== null &&
    REFUSAL_CODES.includes((detail as { code?: unknown }).code as string)
  );
}

/** What a reader can actually do about it.
 *
 *  This used to read "Switch your acting role in the header", which was true
 *  while a control in the header changed the acting role. Under the capability
 *  layer a role is assigned, not chosen, and that control is gone — so the
 *  sentence named an action the reader could not take, which is worse than
 *  naming none. The advice is now the one thing that does move the situation.
 */
const NEXT_ACTION = "Ask an administrator if you need this access.";

/**
 * The sentence, from the role and the action.
 *
 * Falls back to naming neither rather than inventing either. An unknown role
 * or action means this build is older than the server, and a reader is better
 * served by a true vague sentence than a confident wrong one.
 */
export function actorRoleRefusalText(refusal: {
  required_role: string;
  action: string;
}): string {
  const role = ACTOR_ROLE_LABEL[refusal.required_role];
  const phrase = GUARDED_ACTION_PHRASE[refusal.action];

  if (!role) {
    return `You do not have the role this action needs. ${NEXT_ACTION}`;
  }
  if (!phrase) {
    return `Only a ${role} can do this. ${NEXT_ACTION}`;
  }
  return `Only a ${role} can ${phrase}. ${NEXT_ACTION}`;
}
